### PR TITLE
Johtaylo/missing adapter integration interface and InvokeResponse bugs

### PR DESCRIPTION
**Pull request by johnataylor**
- add interface to capture contract between integration pieces and adapter
- make integration add BotFrameworkAdapter as a default rather than hardcoded
- fix InvokeResponse bugs:
1. should return 501
1. should allow null body
- add new debugging and testing Project for WebApi
- further fix up mocks for IServiceCollection collection behavior (necessary because TryAdd style methods uses Linq Any)

We should now add samples for how to provide your own adapter using the out of the box integration pieces (see test projects), in addition to updating the new sample of traditional MVC integration.

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 1]
```diff
@@ -229,15 +231,14 @@ public async Task<InvokeResponse> ProcessActivityAsync(ClaimsIdentity identity,
...
+                    if (activityInvokeResponse == null)
                     {
-                        // ToDo: Trace Here
-                        throw new InvalidOperationException("Bot failed to return a valid 'invokeResponse' activity.");
+                        return new InvokeResponse { Status = (int)HttpStatusCode.NotImplemented };
                     }
                     else
                     {
-                        return (InvokeResponse)invokeResponse.Value;
+                        return (InvokeResponse)activityInvokeResponse.Value;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Why the cast here? Shouldn't the cast on line 234 be changes to: 
`var activityInvokeResponse = context.TurnState.Get<InvokeResponse>(InvokeReponseKey);`

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/BotMessageHandlerBase.cs [hunk 2]
```diff
@@ -61,7 +61,7 @@ public async Task HandleAsync(HttpContext httpContext)
             }
 
             var requestServices = httpContext.RequestServices;
-            var botFrameworkAdapter = requestServices.GetRequiredService<BotFrameworkAdapter>();
+            var adapter = requestServices.GetRequiredService<IAdapterIntegration>();
```
Comment (cleemullins) [inline libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/BotMessageHandlerBase.cs]: Being nitpicky, this isn't an adapter. It'a an integration. :)

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 1 above)
Comment (johnataylor) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: No its actually an Activity on the TurnState. The InvokeResponse is hanging off the Activity. There is no change here - this was intentional.

I was hoping to clarify this confusing contract with the variable name change.

Comment (johnataylor) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: The change here is around returning 501 from the HTTP request. That's the point of this change.

**Comment by johnataylor** [comment]
fixes #1028 #1039

**Pull request merged**